File: tests/07_phases_ompss.dg/fortran/success_task_25.f90
! <testinfo>
! test_generator=config/mercurium-ompss
! </testinfo>

MODULE M
IMPLICIT NONE

TYPE :: T
   INTEGER :: VAR
END TYPE T

TYPE(T) :: X
CONTAINS
    SUBROUTINE SUB1()
        IMPLICIT NONE
        INTERFACE
           !$OMP TARGET DEVICE(SMP) COPY_DEPS
           !$OMP TASK INOUT(X)
           SUBROUTINE SUB2(X)
              IMPORT T
              IMPLICIT NONE
              TYPE(T) :: X
           END SUBROUTINE SUB2
        END INTERFACE

       CALL SUB2(X)
    END SUBROUTINE SUB1
END MODULE M

SUBROUTINE SUB2(X)
   USE M, ONLY : T
   IMPLICIT NONE
   TYPE(T) :: X

   X % VAR = 1
END SUBROUTINE SUB2

PROGRAM MAIN
    USE M
    IMPLICIT NONE
    X % VAR = -1
    CALL SUB1()
    !$OMP TASKWAIT
    IF (X % VAR /= 1) THEN
        STOP 1
    ENDIF
END PROGRAM

File: tests/07_phases_ompss.dg/c/success_task_25.c
/*
<testinfo>
test_generator=(config/mercurium-ompss "config/mercurium-ompss-2 openmp-compatibility")
</testinfo>
*/
#include<assert.h>

int x;
int *v;

#pragma omp task inout(v[x])
void f()
{
    assert(v[x] == 1);
    v[x]--;
}

int main()
{
    x = 0;
    int a = 1;
    v = &a;

    f();

    #pragma omp task inout(v[x])
    {
        assert(v[x] == 0);
        v[x]++;
    }
    #pragma omp taskwait
    assert(v[x] == 1);
    return 0;
}
